compiler/rustc_arena/src/lib.rs

```diff
@@ -109,6 +112,55 @@ impl<T> Default for TypedArena<T> {
     }
 }
 
+trait IterExt<I, T> {
+    fn write_to_arena(iter: I, arena: &TypedArena<T>) -> &mut [T];
+}
+
+impl<T, I> IterExt<I, T> for I
+where
+    I: Iterator<Item = T>,
+{
+    #[inline]
+    default fn write_to_arena(iter: I, arena: &TypedArena<T>) -> &mut [T] {
+        arena.alloc_from_iter_gen(iter)
+    }
+}
+
+impl<T, I> IterExt<I, T> for I
+where
+    I: Iterator<Item = T> + TrustedLen,
+{
+    #[inline]
+    fn write_to_arena(mut iter: I, arena: &TypedArena<T>) -> &mut [T] {
+        let size_hint = iter.size_hint();
+
+        match size_hint {
+            (0, Some(_)) => &mut [],
+            (len, Some(_)) => {
+                // no need to check min == max because of TrustedLen
+                // We know the exact number of elements the iterator will produce here
+                arena.ensure_capacity(len);
+
+                // SAFETY: TrustedLen implementors must ensure they return exactly as man
+                // elements as they tell via `size_hint()`.
+                unsafe {
+                    let start_addr = arena.ptr.get();
+                    let mut mem = start_addr;
+                    // Set the pointer here so that the iterator can also allocate in this arena
+                    arena.ptr.set(mem.add(len));
+                    while let Some(value) = iter.next() {
+                        ptr::write(mem, value);
+                        mem = mem.add(1);
+                    }
+
+                    return slice::from_raw_parts_mut(start_addr, len);
+                }
+            }
+            (_, _) => cold_path(move || arena.alloc_from_iter_gen(iter)),
```

Comment:
I think in this instance we can just panic because if the maximum is `None`, the iterator will never terminate, and thus create an unbounded allocation. We are already guaranteeing that `T` has non-zero size.

Reply:
If I read the docs correctly, `(usize::MAX, None)` just means there are more than `usize::MAX` elements in the iterator. Granted, that means a little bit of memory, but doesn't necessarily diverge.

Reply:
Hmmm good point. But as it needs allocation of at least one byte per iteration run, it'd eventually allocate more than the address space allows. So it'd be invalid one way or another.

Reply:
Well then, this code can be though of as binary size optimization. Jokes aside, I don't mind changing it, and probably a panic could give more information than the OOM killer.